src/com/xilinx/rapidwright/rwroute/RouterHelper.java

```diff
@@ -371,28 +373,56 @@ private static boolean isInvertibleDSPBELPin(BELPin belPin) {
     public static Set<SitePinInst> invertPossibleGndPinsToVccPins(Design design, List<SitePinInst> pins) {
         Net staticNet = design.getGndNet();
         Set<SitePinInst> toInvertPins = new HashSet<>();
-        for (SitePinInst currSitePinInst : pins) {
+        nextSitePin: for (SitePinInst currSitePinInst : pins) {
             if (!currSitePinInst.getNet().equals(staticNet))
                 throw new RuntimeException(currSitePinInst.toString());
-            BELPin[] belPins = currSitePinInst.getSiteInst().getSiteWirePins(currSitePinInst.getName());
-            // DSP or BRAM
-            if (belPins.length == 2) {
-                for (BELPin belPin : belPins) {
-                    if (belPin.isSitePort())    continue;
-                    if (currSitePinInst.getSite().getName().startsWith("RAM")) {
-                        if (!belPin.getBEL().canInvert()) {
-                            continue;
-                        }
-                        if (belPin.getBELName().startsWith("CLK")) {
-                            continue;
-                        }
-                        toInvertPins.add(currSitePinInst);
-                    } else if (currSitePinInst.getSite().getName().startsWith("DSP")) {
-                        if (isInvertibleDSPBELPin(belPin)) {
+            if (currSitePinInst.isLUTInputPin()) {
+                Collection<Cell> connectedCells = DesignTools.getConnectedCells(currSitePinInst);
+                for (Cell cell : connectedCells) {
+                    if (!LUTTools.isCellALUT(cell)) {
+                        continue nextSitePin;
+                    }
+                }
+
+                toInvertPins.add(currSitePinInst);
+
+                for (Cell cell : connectedCells) {
+                    // Find the logical pin name
+                    String physicalPinName = "A" + currSitePinInst.getName().charAt(1);
+                    String logicalPinName = cell.getLogicalPinMapping(physicalPinName);
+
+                    // Get the LUT equation
+                    String lutEquation = LUTTools.getLUTEquation(cell);
+                    assert(lutEquation.contains(logicalPinName));
+
+                    // Compute a new LUT equation with that logical input inverted
+                    String newLutEquation = lutEquation.replace(logicalPinName, "!" + logicalPinName)
+                            // Cancel out double inversions
+                            .replace("!!", "");
```

Comment:
Although not common, Vivado does allow for `~` as an alternative to `!` to specify inversion:
![image](https://github.com/Xilinx/RapidWright/assets/20803522/42d0d726-79a0-42fb-b412-48c4ea01cc9f)

Reply:
My understanding is that the full 2^K LUT mask is stored as the `LUT_INIT` property. This `LUTTools.getLUTEquation()` parses this `LUT_INIT` property into a LUT equation, and it only uses `!` instead of `~` so I think what exists should be okay. Added a comment.

Reply:
You're right, depending on `getLUTEquation()` should prevent any `~` from showing up.